### This Notbook is my personal learning journey with ChatGPT

### 📌 Step 3: Variables (Very Important!)

Variables store information.

#### Exercise 2

Create variables for:

* Your name
* Your age
* Your favorite number

Then print them.

In [7]:
# Exercise 2
name = "M Abid"
age = 25
fav_num = 27

print(name)
print(age)
print(fav_num)

M Abid
25
27


Excellent work, Abid! 👏🔥

You did everything correctly.

#### Exercise 2 Review

```python
name = "M Abid"
age = 25
fav_number = 27

print(name)
print(age)
print(fav_number)
```

✔ Variables created correctly
✔ Correct data types (string + integers)
✔ Correct printing

Very good start

### 🚀 Step 4: Numbers & Math in Python

Python can do math like a calculator.

## Basic Operators

| Symbol | Meaning        |
| ------ | -------------- |
| `+`    | Addition       |
| `-`    | Subtraction    |
| `*`    | Multiplication |
| `/`    | Division       |
| `**`   | Power          |
| `%`    | Remainder      |


#### 🧠 Bonus Challenge (Important for problem solving)

Try this:

```python
num = 7
```

Print:

* Double of the number
* Triple of the number
* Square of the number

In [12]:
# Exercise 3
x = 15
y = 4

print(x + y)
print(x - y)
print(x * y)
print(x / y)
print(x % y)

19
11
60
3.75
3


In [9]:
# Bonus Challenge
num = 7

print(num * 2)
print(num * 3)
print(num ** 2)

14
21
49


#### Exercise 3 Review

```python
x = 15
y = 4
print(x + y)
print(x - y)
print(x * y)
print(x / y)
print(x % y)
```

✔ Addition → 19
✔ Subtraction → 11
✔ Multiplication → 60
✔ Division → 3.75
✔ Remainder → 3

Perfect 👌

#### Bonus Review

```python
num = 7
print(num * 2)
print(num * 3)
print(num ** 2)
```

✔ Double → 14
✔ Triple → 21
✔ Square → 49

Excellent thinking

### 🚀 Step 5: User Input (Now it gets interesting!)

Now your program can **talk to the user**.

⚠ Important:
`input()` always gives you TEXT (string).

If you want a number, you must convert it:


#### Exercise 4

1. Ask the user for their name.
2. Ask the user for their age.
3. Print:

   ```
   Hello Abid, next year you will be 26
   ```

(Hint: You need `int()` for age.)

#### 🎯 Mini Challenge (Think a little!)

Make a small program that:

* Asks for two numbers
* Prints their sum

Example output:

```
Enter first number: 5
Enter second number: 3
The sum is 8
```

In [13]:
# Exercise 4
name = input("What is your name? ")
age = int(input("What is your age? "))

print("Hello", name, "next year you will be", age + 1)

Hello M Abid next year you will be 26


In [15]:
# Mini Challenge
first_number = int(input("Enter first number: "))
second_number = int(input("Enter second number: "))
total = first_number + second_number

print("The sum is:", total)

The sum is: 8


Let’s review carefully:

#### Exercise 4 Review

```python
name = input("What is your name? ")
age = int(input("Enter you age: "))
print("Hello", name, "next year you will be", age + 1)
```

🟢 Very good:

* Correct use of `input()`
* Correct use of `int()`
* Correct printing

#### Mini Challenge Review

```python
first_number = int(input("Enter first number: "))
second_number = int(input("Enter second number: "))
total = first_number + second_number
print("The sum is ", total)
```

✔ Perfect logic
✔ Correct conversion
✔ Correct addition

### 🚀 Step 6: If Statements (Now we enter real programming)

This is where logic begins.

Example:

```python
age = 18

if age >= 18:
    print("You are an adult")
```

Structure:

```python
if condition:
    code
```

⚠ Indentation (spaces) is VERY important in Python.

#### Exercise 5

Create a program that:

1. Asks the user for their age.
2. If age is 18 or older → print `"You can vote"`
3. Otherwise → print `"You cannot vote"`

(Hint: You need `if` and `else`)

#### Bonus Challenge

Ask the user for a number.
If the number is even → print `"Even number"`
If the number is odd → print `"Odd number"`

(Hint: Use `%`)

In [3]:
# Exercise 5
age = int(input("What is your age? "))

if age >= 18:
    print("You can vote")
else:
    print("You cannot vote")

You can vote


In [7]:
# Bouns Challenge
user_num = int(input("Enter a number: "))

if user_num %2 == 0:
    print("Odd number")
else:
    print("Even number")

Odd number


Nice work again, Abid 👏🔥
Your logic is getting stronger.

#### Exercise 5 (Voting)

```python
age = int(input("Enter your age: "))
if age >= 18:
    print("You can vote")
else:
    print("You cannot vote")
```

✔ Correct condition
✔ Correct `if/else` structure
✔ Proper indentation

Perfect ✅

#### 🔎 Bonus Challenge Review (Even / Odd)

Your code:

```python
user_num = int(input("Enter your number: "))
if user_num %2 == 0:
    print("Odd number")
else:
    print("Even number")
```

You made a **small logical mistake**:

`user_num % 2 == 0` means the number is EVEN.

But you printed `"Odd number"`.

#### Correct Version:

```python
user_num = int(input("Enter your number: "))
if user_num % 2 == 0:
    print("Even number")
else:
    print("Odd number")
```

#### Why?

* `%` gives remainder
* Even numbers divided by 2 have remainder 0
* Odd numbers have remainder 1

Example:

* `8 % 2 = 0` → Even
* `7 % 2 = 1` → Odd


### 🚀 Step 7: Multiple Conditions (elif)

Now we level up again.

#### Exercise 6

Create a program that:

1. Asks the user for a number.
2. If number is:

   * Greater than 0 → print `"Positive"`
   * Less than 0 → print `"Negative"`
   * Equal to 0 → print `"Zero"`

Use `if`, `elif`, and `else`.

#### 🎯 Mini Challenge (Logic Builder)

Make a small calculator that:

* Asks for two numbers
* Asks for an operator (`+`, `-`, `*`, `/`)
* Performs the correct calculation using `if/elif`

Example:

```
Enter first number: 5
Enter operator: *
Enter second number: 3
Result: 15
```

In [ ]:
# Exercise 6
number = int(input("Enter your number "))
if number > 0:
    print("Possitive number")
elif number < 0:
    print("Negative number")
else:
    print("Zero number")

Negative number


In [23]:
# Mini Challenge
# Basic Math Operators (+ - * /)

first_number = int(input("Enter first number: "))
second_number = int(input("Enter second number"))
operator = input("Chose one operator: ")

if operator == "+":
    print("Result:", first_number + second_number)
elif operator == "-":
    print("Result:", first_number - second_number)
elif operator == "*":
    print("Result:", first_number * second_number)
elif operator == "/":
    print("Result:", first_number / second_number)
else:
    print("Invaled operator")

Invaled operator


Excellent work again, Abid! You’re clearly understanding the concepts well.

#### Exercise 6 Review (Positive / Negative / Zero)

Your code:

```python
number = int(input("Enter your number: "))

if number > 0:
    print("Positive number")
elif number < 0:
    print("Negative number")
else:
    print("Zero number")
```

✔ Perfect logic
✔ Correct use of `if / elif / else`
✔ Correct conditions

This is exactly how a programmer would write it. Great job 👌

#### Mini Calculator Review

Your code:

```python
first_number = int(input("Enter first number: "))
operator = input("Enter single operator: ") # Operator (+ , - , * , /)
second_number = int(input("Enter second number: "))

if operator == "+":
    print("Result:", first_number + second_number)
elif operator == "-":
    print("Result:", first_number - second_number)
elif operator == "*":
    print("Result:", first_number * second_number)
elif operator == "/":
    print("Result:", first_number / second_number)
else:
    print("Invalid operator")
```

✔ Correct use of `if / elif / else`
✔ Correct comparisons
✔ Correct math operations

Very good logic 👏

### 🚀 Step 8: Loops (Very Powerful)

Loops let a program **repeat actions automatically**.

#### Exercise 7

Write a program that prints numbers **1 to 10** using a loop.

Expected output:

```
1
2
3
4
5
6
7
8
9
10
```

#### 🎯 Bonus Challenge

Print the **multiplication table of 5**.

Example:

```
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
...
5 x 10 = 50
```

Hint:

```
for i in range(1, 11):
```


In [27]:
# Exercise 7
for i in range(1,11):
    print(i)

1
2
3
4
5
6
7
8
9
10


In [33]:
# Bouns Challenge
for i in range(1, 11):
    print(5, "x", i, "=", i * 5)

5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50


Great work again, Abid! 👏🔥 You’re doing **exactly what a good beginner should do: write code and practice**.

#### Exercise 7 Review

Your code:

```python
for i in range(1,11):
    print(i)
```

✔ Correct loop
✔ Correct range
✔ Correct output

This will print numbers **1 → 10**, exactly as required. Perfect ✅

#### Bonus Challenge Review (Multiplication Table)

Your code:

```python
for i in range(1, 11):
    print(5, "x", i, "=", i * 5)
```

✔ Correct loop
✔ Correct multiplication logic
✔ Correct formatting

Output will look like:

```
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
...
5 x 10 = 50
```

Excellent 👍


### 🚀 Step 9: `while` Loops

A **while loop runs as long as a condition is true**.

### Important part

```python
i += 1
```

This increases the number each loop.
Without it, the loop would run **forever** (infinite loop).

#### Exercise 8

Write a program that prints numbers **1 to 5 using a `while` loop**.

Expected output:

```
1
2
3
4
5
```

#### 🎯 Mini Challenge (Loop + Logic)

Make a program that:

1. Starts with number `10`
2. Prints numbers **10 → 1**
3. Then prints `"Blast off!"`

Expected output:

```
10
9
8
7
6
5
4
3
2
1
Blast off!
```

In [40]:
# Exercise 8
i = 1

while i <= 5:
    print(i)
    i += 1

1
2
3
4
5


In [36]:
# Mini Challenge
for i in range(10, 0, -1):
    print(i)
    
print("Blast off!")

10
9
8
7
6
5
4
3
2
1
Blast off!


Excellent, Abid! 🔥 You’re really getting comfortable with Python now.

#### Exercise 8 Review (`while` loop)

Your code:

```python
i = 1

while i <= 5:
    print(i)
    i += 1
```

✔ Correct `while` condition
✔ Correct increment (`i += 1`)
✔ Correct output 1 → 5

Perfect solution 👌

#### Mini Challenge Review

Your code:

```python
for i in range(10, 0, -1):
    print(i)

print("Blast off!")
```

✔ Correct countdown
✔ Correct use of `range(start, stop, step)`
✔ Correct step `-1`
✔ Correct final message

Output will be:

```
10
9
8
7
6
5
4
3
2
1
Blast off!
```

Very nice 👍

Even though the challenge was about loops generally, using a **`for` loop** here is actually **cleaner than `while`**, so this is good programming practice.

### 🚀 Step 10: Lists (Super Important)

Lists store **multiple values in one variable**.

Example:

```python
numbers = [1, 2, 3, 4, 5]
```

Another example:

```python
fruits = ["apple", "banana", "orange"]
```

You can access items by **index**.

```python
fruits = ["apple", "banana", "orange"]

print(fruits[0])
print(fruits[1])
```

Output:

```
apple
banana
```

Important rule:

```
Index starts at 0
```

| Index | Value  |
| ----- | ------ |
| 0     | apple  |
| 1     | banana |
| 2     | orange |


#### Exercise 9

Create a list of **5 favorite foods**.

Print:

1. The whole list
2. The first food
3. The last food

Hint:

```python
foods[-1]
```

gets the **last item**.


#### Bonus Challenge

Create a list:

```python
numbers = [10, 20, 30, 40, 50]
```

Write a loop that prints:

```
10
20
30
40
50
```

(Hint: loop through the list)

In [45]:
# Exercise 9
foods = ["Meal of Wheat", "Rice", "Meat of Goat", "Fresh Fish", "Butter with honey"]
print(foods)
print(foods[0])
print("The last food:", foods[-1])

['Meal of Wheat', 'Rice', 'Meat of Goat', 'Fresh Fish', 'Butter with honey']
Meal of Wheat
The last food: Butter with honey


In [46]:
# Bonus Challenge
numbers = [10, 20, 30, 40, 50]

for i in numbers:
    print(i)

10
20
30
40
50


Great job again, Abid! 👏🔥 Your Python skills are clearly improving.

#### Exercise 9 Review

Your code:

```python
foods = ["Meal of Wheat", "Rice", "Meat of Goat", "Fresh Fish", "Butter with honey"]
print(foods)
print(foods[0])
print("The last food:", foods[-1])
```

✔ Correct list creation
✔ Correct printing of the whole list
✔ Correct first element `foods[0]`
✔ Correct last element `foods[-1]`

Perfect solution 👌

Example output:

```
['Meal of Wheat', 'Rice', 'Meat of Goat', 'Fresh Fish', 'Butter with honey']
Meal of Wheat
The last food: Butter with honey
```

#### Bonus Challenge Review

Your code:

```python
numbers = [10, 20, 30, 40, 50]

for i in numbers:
    print(i)
```

✔ Correct loop through list
✔ Correct syntax
✔ Clean and professional solution

Output:

```
10
20
30
40
50
```

This is exactly how Python programmers loop through lists. Nice work 💪

### 🚀 Step 11: List Operations

Lists can **grow and change**.

#### List length

```python
print(len(fruits))
```

#### Exercise 10

Create this list:

```python
animals = ["cat", "dog", "cow"]
```

Then:

1️⃣ Add `"goat"` to the list
2️⃣ Remove `"dog"`
3️⃣ Print the final list
4️⃣ Print the **length of the list**

#### 🎯 Small Practice (Very Important)

Try this yourself.

Make a program that:

1️⃣ Creates an empty list
2️⃣ Asks the user for **5 numbers**
3️⃣ Adds them to the list
4️⃣ Prints the list

Example:

```
Enter number: 4
Enter number: 7
Enter number: 2
Enter number: 9
Enter number: 1

Numbers: [4, 7, 2, 9, 1]
```


In [65]:
# Exercise 10
animals = ["cat", "dog", "cow"]
animals.append("goat")
animals.remove("dog")

print("The final list:", animals)
print(len(animals))

The final list: ['cat', 'cow', 'goat']
3


In [66]:
# Small Practice
numbers = []

for i in range(5):
    number = int(input("Enter your number: "))
    numbers.append(number)

print("Your provided all numbers are: ",numbers)

Your provided all numbers are:  [99, 12, 32, 76, 100]


#### Exercise 10 Review

Your code:

```python
animals = ["cat", "dog", "cow"]
animals.append("goat")
animals.remove("dog")

print("The final list:", animals)
print(len(animals))
```

✔ Correct `append()`
✔ Correct `remove()`
✔ Correct `len()`

Final list will be:

```
['cat', 'cow', 'goat']
```

Perfect solution 👌

#### Small Practice Review

Your small practice solution:

```python
numbers = []

for i in range(5):
    number = int(input("Enter your number: "))
    numbers.append(number)

print("Your provided all numbers are: ",numbers)
```

This is **100% correct**. ✅
Many beginners struggle with this — but you solved it.

### 🧠 Now Let’s Continue Your Training

Next we learn something **VERY important**:

#### Functions

Functions are how real programs are built.

Example:

```python
def greet():
    print("Hello Abid")

greet()
```

Output:

```
Hello Abid
```

#### Exercise 11

Create a function:

```python
def say_hello(name):
```

That prints:

```
Hello Abid
```

Call it like:

```
say_hello("Abid")
```

#### 🎯 Small Challenge

Create a function:

```
add_numbers(a, b)
```

It should print the sum.

Example:

```
add_numbers(5,3)
```

Output:

```
8
```


In [12]:
def greet():
    print("Assalam o Alaikum M Abid")

greet()

Assalam o Alaikum M Abid


In [13]:
# Exercise 11
def say_hello(name):
    print("Hello", name)

say_hello("Abid")

Hello Abid


In [8]:
# Small Challenge
def add_numbers(a, b):
    print(a + b)

add_numbers(5,3)

8


Abid, first let me say: **your solutions are correct again.** You’re progressing steadily, and that consistency is exactly what leads to real skills. 💪

#### Exercise 11 Review

Your code:

```python
def say_hello(name):
    print("Hello", name)

say_hello("Abid")
```

✔ Correct function definition
✔ Correct parameter usage
✔ Correct function call

Output:

```
Hello Abid
```

Perfect. ✅

#### Small Challenge Review

Your code:

```python
def add_numbers(a, b):
    print(a + b)

add_numbers(5,3)
```

✔ Correct function
✔ Correct parameters
✔ Correct logic

Output:

```
8
```

Excellent. 👌

Small professional improvement later (not required now):

Sometimes functions **return values** instead of printing them:

```python
def add_numbers(a, b):
    return a + b

result = add_numbers(5,3)
print(result)
```

But what you wrote is **completely correct for your level**.

In [11]:
def multiply_numbers(a, b):
    return a * b
result = multiply_numbers(8, 3)
print(result)

24


### 🚀 Your Next Lesson (Important)

Now we learn **Dictionaries**.

Lists store values.

Dictionaries store **key → value pairs**.

#### Exercise 12

Create this dictionary:

```python
person = {
    "name": "Abid",
    "age": 25,
    "country": "Pakistan"
}
```

Print:

```
Name:
Age:
Country:
```

Using dictionary keys.

#### 🎯 Small Challenge

Ask the user for:

* name
* age
* favorite programming language

Store them in a dictionary and print the dictionary.

In [38]:
# Exercise 12
person = {
    "name": "M Abid",
    "age": 25,
    "country": "Pakistan"
}

print(person["name"])
print(person["age"])
print(person["country"])

M Abid
25
Pakistan


In [37]:
# Small Challenge
name = input("What is your name? ")
age = int(input("What is your age? "))
fav_prg_lng = input("What is your favorite programming language? ")

new_dict = {
    "name": name,
    "age": age,
    "fav_prg_lng": fav_prg_lng
}

print("Name:", new_dict["name"])
print("Age:", new_dict["age"])
print("Favorite Language:", new_dict["fav_prg_lng"])

Name: M Abid
Age: 25
Favorite Language: Python


Hello Abid! 👋
I’m doing well—thanks for asking. And I’m proud to see you coming back with your progress. Let’s review your work like a real coach would. 💪📚

#### Exercise 12 Review

Your code:

```python
person = {
    "name": "M Abid",
    "age": 25,
    "country": "Pakistan"
}

print(person["name"])
print(person["age"])
print(person["country"])
```

✔ Dictionary created correctly
✔ Keys used correctly
✔ Values printed correctly

Output will be:

```
M Abid
25
Pakistan
```

Perfect work. ✅

#### Small Challenge Review

Your code:

```python
name = input("What is your name? ")
age = int(input("What is your age? "))
fav_prg_lng = input("What is your favorite programming language? ")

new_dict = {
    "name": name,
    "age": age,
    "fav_prg_lng": fav_prg_lng
}

print("Name:", new_dict["name"])
print("Age:", new_dict["age"])
print("Favorite Language:", new_dict["fav_prg_lng"])
```

✔ Good use of `input()`
✔ Correct dictionary structure
✔ Correct key access with formating

This shows you **understand dictionaries well already**. 👍


### Important Concept: Dictionaries + Loops

A powerful thing programmers do is **loop through dictionaries**.

Example:

```python
person = {
    "name": "Abid",
    "age": 25,
    "country": "Pakistan"
}

for key, value in person.items():
    print(key, ":", value)
```

Output:

```
name : Abid
age : 25
country : Pakistan
```


#### 🚀 Exercise 13

Create a dictionary:

```python
student = {
    "name": "Abid",
    "age": 25,
    "course": "Python"
}
```

Add a **new key**:

```
"city": "Islamabad"
```

Then print the full dictionary.

Hint:

```python
student["city"] = "Islamabad"
```

#### 🎯 Challenge (Important Logic Practice)

Create a **mini contact book**.

Steps:

1️⃣ Create an empty dictionary

```
contacts = {}
```

2️⃣ Ask user for:

* name
* phone number

3️⃣ Store it in dictionary like this:

```
contacts[name] = phone
```

4️⃣ Print the contacts dictionary.

Example run:

```
Enter name: Ali
Enter phone: 12345

{'Ali': '12345'}
```

#### 🧭 Coaching Advice for You

Abid, your learning pattern is **very good**:

✔ You attempt every exercise
✔ You write code yourself
✔ You ask questions honestly

This is **exactly how developers are made**.

If you continue like this **3–4 months**, you will be far ahead of many beginners.

In [39]:
# Exercise 13
student = {
    "name": "Abid",
    "age": 25,
    "course": "Python"
}

student["city"] = "Islamabad"
print(student)

{'name': 'Abid', 'age': 25, 'course': 'Python', 'city': 'Islamabad'}


In [42]:
# Challenge
contacts = {}

name = input("What is your name? ")
phone_number = int(input("Enter your phone numbere here "))

contacts["name"] = name
contacts["phone number"] = phone_number

print(contacts)
print("Name:",contacts["name"])
print("Phone number:",contacts["phone number"])

{'name': 'M Abid', 'phone number': 1234567}
Name: M Abid
Phone number: 1234567


Abid, great to see your consistency. Studying **4–5 hours daily** and practicing in **VS Code** is an excellent habit. That level of discipline can absolutely change your situation if you keep it up. 💪

Let’s review your code carefully.

#### Exercise 13 Review

Your code:

```python
student = {
    "name": "Abid",
    "age": 25,
    "course": "Python"
}

student["city"] = "Islamabad"
print(student)
```

✔ Correct dictionary
✔ Correct way to add a new key
✔ Correct output

Example output:

```
{'name': 'Abid', 'age': 25, 'course': 'Python', 'city': 'Islamabad'}
```

Perfect solution. ✅

#### 🔎 Challenge Review (Contacts)

Your code works, but there is a **small logical improvement**.

Your version:

```python
contacts = {}

name = input("What is your name? ")
phone_number = int(input("Enter your phone numbere here "))

contacts["name"] = name
contacts["phone number"] = phone_number
```

This creates a dictionary like:

```
{
 "name": "Abid",
 "phone number": 12345
}
```

But the goal was to store **name → phone number** like a contact book.

Better structure:

```
{
 "Abid": "12345"
}
```

#### Improved Solution

```python
contacts = {}

name = input("Enter name: ")
phone = input("Enter phone number: ")

contacts[name] = phone

print(contacts)
```

Example run:

```
Enter name: Ali
Enter phone number: 12345

{'Ali': '12345'}
```

Now the **name becomes the key**.

This is how real contact dictionaries work.


#### 🎯 Small Exercise For You

Modify your program to:

1️⃣ Store 5 contacts
2️⃣ Ask user **which name they want to search**
3️⃣ Print that phone number

Example:

```
Enter name to search: Abid
Phone number: 444
```


In [113]:
# Exercise 13
phone_diary = {}

for i in range(5):
    name = input("Enter your name: ")
    cell = input("Enter your phone number: ")
    phone_diary[name] = cell

search = input("Write a name to find phone number: ")

if search in phone_diary:
    print("Your required phone number is: ",phone_diary[search])
else:
    print("Invalid user name")

print() # It's for empty space between below and above prints

for key, value in phone_diary.items():
    print(key, ":", value)

Your required phone number is:  444

Ali : 111
Sara : 222
Bob : 333
Abid : 444
David : 555


#### ✔ Things you did correctly

1. **Dictionary creation**

```python
phone_diary = {}
```

2. **Loop to add multiple contacts**

```python
for i in range(5):
```

3. **Key → value storage**

```python
phone_diary[name] = cell
```

4. **Search logic**

```python
if search in phone_diary:
```

5. **Fetching the value**

```python
phone_diary[search]
```

This is **exactly how dictionary lookup works**.

#### Important Logic You Just Learned

You combined **4 core programming skills**:

```
loops
input
dictionary
condition
```

This combination is used **everywhere**:

* login systems
* chatbots
* AI agents
* databases
* APIs

You are learning **real programming structure now**.

### 🚀 Next Concept: File Handling

Right now your programs **lose data when they stop**.

Example:

```
contacts disappear when program ends
```

File handling allows us to **save data permanently**.

Example file:

```
contacts.txt
```

#### Example

```python
file = open("notes.txt", "w")
file.write("I am learning Python")
file.close()
```

This creates a file.

#### ✅ Exercise 14

Write a program that:

1️⃣ Asks the user to enter a note
2️⃣ Saves it to a file called `"notes.txt"`

Example:

```
Enter a note: Study Python today
```

After running, `notes.txt` should contain:

```
Study Python today
```

#### 🎯 Small Challenge

Make a **task saver program**.

Steps:

1️⃣ Ask user for a task
2️⃣ Save it to `"tasks.txt"`
3️⃣ Each new task should be added to the file

Hint: use `"a"` mode instead of `"w"`.

In [1]:
file = open("notes.txt", "w")
file.write("I am learning Python")
file.close()

In [2]:
# Exercise 14
file = open("notes.txt", "w")
note = input("Enert a note: ")
file.write(note)
file.close()

#### Exercise 14 Review (File Writing)

Your code:

```python
file = open("notes.txt", "w")
note = input("Enert a note: ")
file.write(note)
file.close()
```

✔ Correct file mode `"w"`
✔ Correct `write()`
✔ Correct `close()`

Output:

```
Study Python today
```

Perfect.

Small improvement professionals use:

```python
with open("notes.txt", "w") as file:
    note = input("Enter a note: ")
    file.write(note)
```

`with open` automatically closes the file.

In [3]:
# Small Challenge
file = open("tasks.txt", "a")
task = input("Enter your task")
file.write(task)
file.close()

#### Small Challenge Review (Append Mode)

Your code:

```python
file = open("tasks.txt", "a")
task = input("Enter your task")
file.write(task)
file.close()
```

Works correctly.

But better formatting:

```python
file.write(task + "\n")
```

Otherwise tasks may appear in one line.

In [24]:
# Challenge (Your Firs Real Mini Project)
my_book = {}

for i in range(5):
    name = input("Enter name: ")
    phone = input("Enter phone: ")
    my_book[name] = phone

    file = open("contacts.txt", "a")
    file.write(name + " : " + phone + "\n")
    file.close()

search = input("Enter name to find phone number: ")

if search in my_book:
    print("Phone number:", my_book[search])
else:
    print("Invalid user name: ")

print()

for name, phone in my_book.items():
    print(name, ":", phone)

Invalid user name: 

Ali : 111
Sara : 222
David : 333
Abid : 444
Bob : 555


#### Mini Project Review (Phone Diary)

Your program is actually **very good**.

You combined:

* dictionary
* loop
* file saving
* search
* iteration

That’s exactly how beginner projects look.

Your file output:

```
Ali : 111
Sara : 222
David : 333
Abid : 444
Bob : 555
```

Correct.

#### Why You Saw "Invalid user name"

Your program works. That message appears only if the user searches for a name **not inside the dictionary**.

Example:

Dictionary contains:

```
Ali
Sara
David
Abid
Bob
```

If you search:

```
John
```

Output:

```
Invalid user name
```

So your program is behaving **correctly**.